modules/comms/src/Comms.ts

```diff
@@ -0,0 +1,143 @@
+    this.smsService = Sms.getInstance(this.grpcSdk);
+    // this.emailService = Email.getInstance(this.grpcSdk);
+    // this.pushService = Sms.getInstance(this.grpcSdk);
+    this.service = {
+      protoPath: path.resolve(__dirname, 'comms.proto'),
+      protoDescription: 'comms',
+      functions: {
+        Comms: {
+          featureAvailable: this.featureAvailable.bind(this),
+        },
+        Sms: {
+          ...(this.smsService.rpcFunctions as { [key: string]: ServiceFunction }),
+        },
+        Email: {
+          ...(this.emailService.rpcFunctions as { [key: string]: ServiceFunction }),
+        },
+        PushNotifications: {
+          ...(this.pushService.rpcFunctions as { [key: string]: ServiceFunction }),
+        },
+      },
+    };
+  }
+
+  async preConfig(config: Config) {
+    let modifiedConfig = config;
+    if (!config.migrationComplete) {
+      const sms = await this.grpcSdk.config.get('sms');
+      const email = await this.grpcSdk.config.get('email');
+      const push = await this.grpcSdk.config.get('pushNotifications');
+      modifiedConfig = {
+        ...config,
+        sms: {
+          ...sms,
+        },
+        email: {
+          ...email,
+        },
+        push: {
+          ...push,
+        },
+        migrationComplete: true,
+      };
+    }
+
+    modifiedConfig =
+      (await this.smsService.preConfig?.(modifiedConfig)) ?? modifiedConfig;
+    modifiedConfig =
+      (await this.emailService.preConfig?.(modifiedConfig)) ?? modifiedConfig;
+    modifiedConfig =
+      (await this.pushService.preConfig?.(modifiedConfig)) ?? modifiedConfig;
+    return modifiedConfig;
+  }
+
+  async onServerStart() {
+    await this.smsService.onServerStart?.();
+    await this.emailService.onServerStart?.();
+    await this.pushService.onServerStart?.();
+  }
+
+  async onConfig() {
+    await this.smsService.onConfig?.();
+    await this.emailService.onConfig?.();
+    await this.pushService.onConfig?.();
+    let oneHealthy = false;
+    if (this.smsService.health === HealthCheckStatus.SERVING) {
+      oneHealthy = true;
+    }
+    if (this.emailService.health === HealthCheckStatus.SERVING) {
+      oneHealthy = true;
+    }
+    if (this.pushService.health === HealthCheckStatus.SERVING) {
+      oneHealthy = true;
+    }
+    if (oneHealthy) {
+      this.updateHealth(HealthCheckStatus.SERVING);
+    } else {
+      this.updateHealth(HealthCheckStatus.NOT_SERVING);
+    }
+    await AdminHandlers.getInstance().registerAdminRoutes();
+
+    this.grpcSdk
+      .waitForExistence('router')
+      .then(() => ClientRouteHandlers.getInstance().registerRoutes())
+      .catch(e => {
+        ConduitGrpcSdk.Logger.error(e.message);
+      });
+  }
+
+  async initializeMetrics() {
+    await this.smsService.initializeMetrics?.();
+    await this.emailService.initializeMetrics?.();
+    await this.pushService.initializeMetrics?.();
+  }
+
+  // gRPC Service
+  async featureAvailable(
```

Comment:
Delivering this RPC method goes against everything I know to be right and true and I will sooner lay you into this barren earth, than entertain your folly for a moment longer. [[1](https://www.youtube.com/watch?v=y8OnoxKotPQ)]

Jokes aside, this pointlessly breaks [gRPC Health Checking Protocol v1](https://github.com/grpc/grpc/blob/master/doc/health-checking.md):

> The client can optionally set the service name it wants to query for health status.
>
> The server should register all the services manually and set the individual status, including an empty service name and its status. For each request received, if the service name can be found in the registry, a response must be sent back with an OK status and the status field should be set to SERVING or NOT_SERVING accordingly. If the service name is not registered, the server returns a NOT_FOUND GRPC status.
>
> The server should use an empty string as the key for server's overall health status, so that a client not interested in a specific service can query the server's status with an empty request. The server can just do exact matching of the service name without support of any kind of wildcard matching. However, the service owner has the freedom to implement more complicated matching semantics that both the client and server agree upon.

``` proto
message HealthCheckRequest {
  string service = 1;
}
```

All that's needed for compatibility to be preserved is for `Comms` to integrate this pattern matching in the context of health `check()`/`watch()`.

While`ConduitServiceModule` doesn't currently explicitly accommodate this functionality, merely overriding the base methods wouldn't cause any additional breaking changes, nor overcomplicate the `module-tools` API for single service consumers.
Here's the current impl for reference:

``` ts
healthCheck(
  call: GrpcRequest<HealthCheckRequest>,
  callback: GrpcCallback<HealthCheckResponse>,
) {
  const service = call.request.service.substring(call.request.service.indexOf('.') + 1);
  if (service && service !== this._serviceName) {
    callback(null, {
      status:
        HealthCheckStatus.SERVICE_UNKNOWN as unknown as HealthCheckResponse_ServingStatus,
    });
  } else {
    if (!this._registered) {
      return callback(null, {
        status: HealthCheckResponse_ServingStatus.UNKNOWN,
      });
    }
    callback(null, {
      status: this._serviceHealthState as unknown as HealthCheckResponse_ServingStatus,
    });
  }
}

healthWatch(call: ServerWritableStream<HealthCheckRequest, HealthCheckResponse>) {
  const service = call.request.service.substring(call.request.service.indexOf('.') + 1);
  if (service && service !== this._serviceName) {
    call.write({
      status:
        HealthCheckStatus.SERVICE_UNKNOWN as unknown as HealthCheckResponse_ServingStatus,
    });
  } else {
    this.events.on(
      `grpc-health-change:${this._serviceName}`,
      (status: HealthCheckStatus) => {
        call.write({
          status: status as unknown as HealthCheckResponse_ServingStatus,
        });
      },
    );
  }
}
```

Reply:
@kon14 I agree, but this can't be part of this PR to be honest. I see 2 possibilities, either we make changes for health-checking utilities in another PR, merge it and apply changes here as well, or we merge this as-is, change the health checking and apply changes afterwards. 

I prefer no2 because I am doing a re-work of health checking from the core side, in order for it to work properly with dns names on platforms like k8s or docker compose when DNS is involved, so in the same PR I can tackle sdk/module-tools checks for multi-service modules as well. If you agree with that, open it as an issue and I'll add it to the roadmap (I'm preparing the 0.17 roadmap project)